raiden/network/proxies/token_network.py

```diff
@@ -1341,78 +1393,166 @@ def update_transfer(
     def unlock(
         self,
         channel_identifier: ChannelID,
-        participant: Address,
-        partner: Address,
-        merkle_tree_leaves: Optional[MerkleTreeLeaves],
+        sender: Address,
+        receiver: Address,
+        merkle_tree_locks: MerkleTreeLeaves,
+        given_block_identifier: BlockSpecification,
     ):
         log_details = {
             "token_network": pex(self.address),
             "node": pex(self.node_address),
-            "partner": pex(partner),
-            "participant": pex(participant),
-            "merkle_tree_leaves": merkle_tree_leaves,
+            "sender": pex(sender),
+            "receiver": pex(receiver),
+            "merkle_tree_locks": merkle_tree_locks,
         }
 
-        if merkle_tree_leaves is None or not merkle_tree_leaves:
-            log.info("skipping unlock, tree is empty", **log_details)
-            return
+        if not merkle_tree_locks:
+            raise ValueError("unlock cannot be done without merkle_tree_leaves")
+
+        leaves = cast(List[Keccak256], [lock.lockhash for lock in merkle_tree_locks])
+        merkle_tree = MerkleTreeState(compute_layers(leaves))
+
+        # Check the preconditions for calling unlock at the time the event was
+        # emitted.
+        try:
+            channel_onchain_detail = self._detail_channel(
+                participant1=sender,
+                participant2=receiver,
+                block_identifier=given_block_identifier,
+                channel_identifier=channel_identifier,
+            )
+            sender_details = self._detail_participant(
```

Comment:
`detail_channel` already fetches `detail_participant`... is this double fetching done here to make sure that no race condition occurs? If so, i am not sure about the reasoning because i think the execution latency is not going to be large enough for any change on the on-chain details to happen.

Reply:
> detail_channel already fetches detail_participant

Not really. [`_detail_channel`](https://github.com/raiden-network/raiden/blob/2f8f67cb3a2226e2c8518a21334fb15618af4e58/raiden/network/proxies/token_network.py#L391-L429) returns a [`ChannelData`](https://github.com/raiden-network/raiden/blob/2f8f67cb3a2226e2c8518a21334fb15618af4e58/raiden/network/proxies/token_network.py#L118-L121), and [`_detail_participant`](https://github.com/raiden-network/raiden/blob/2f8f67cb3a2226e2c8518a21334fb15618af4e58/raiden/network/proxies/token_network.py#L364-L389) returns [`ParticipantDetails`](https://github.com/raiden-network/raiden/blob/2f8f67cb3a2226e2c8518a21334fb15618af4e58/raiden/network/proxies/token_network.py#L124-L132). 

The first only uses `getChannelInfo` and the latter `getChannelParticipantInfo`, which return a disjoint set of values.